File: submission/12017800/code/02_basic_coop/differentialEvolution.cpp
#include "differentialEvolution.h"

#include <cfloat>
#include <cstddef>
#include <cmath>
#include <thread>
#include <unistd.h>

using namespace std;

volatile BestIndividual* DifferentialEvolution::bestIndividual = NULL;
volatile int DifferentialEvolution::numIterations = 0;

DifferentialEvolution::DifferentialEvolution(Crossover* crossover, OptimizationFunction* optimizationFunction, int numDimensions, int startCellIndex, int endCellIndex) {

	if (bestIndividual == NULL) {
		lock.lock();
		if (bestIndividual == NULL) {
			bestIndividual = new BestIndividual(numDimensions, optimizationFunction);
		}
		lock.unlock();
	}

	this->startCellIndex = startCellIndex;
	this->endCellIndex = endCellIndex;
	this->crossover = crossover;
	this->optimizationFunction = optimizationFunction;

	for (int i = 0; i < populationSize; i++) {
		Individual individual(endCellIndex - startCellIndex, optimizationFunction->getMin(), optimizationFunction->getMax());
		individuals.push_back(individual);
	}
}

DifferentialEvolution::~DifferentialEvolution() {
	if (bestIndividual != NULL) {
		lock.lock();
		if (bestIndividual != NULL) {
			delete bestIndividual;
			bestIndividual = NULL;
		}
		lock.unlock();
	}
}

Individual* DifferentialEvolution::optimize(bool showProgress) {
	int iterationCounter = 0;
	int maxIterations = 5000;
	int noImprovementIterationCounter = 0;
	int maxNoImprovementIterations = 500;
	long double bestFitness = FLT_MAX;

	while (iterationCounter < maxIterations && !isinf(bestFitness)) { // && noImprovementIterationCounter < maxNoImprovementIterations
		long double newBest = createNextGeneration();

		if (newBest >= bestFitness) {
			noImprovementIterationCounter++;
		}
		else {
			noImprovementIterationCounter = 0;
			bestFitness = newBest;
		}

		iterationCounter++;
		
		if (showProgress && iterationCounter % 100 == 0) {
			cout << "\rAfter " << iterationCounter << " iterations, the best individual's fitness is: " << bestFitness << "                    " << flush;
			// usleep(100000); // sleep 1 sec: 1 * 1000 * 1000
		}
	}

	numIterations = iterationCounter;

	return (Individual*)bestIndividual;
}

float DifferentialEvolution::createNextGeneration() {
	vector<Individual> newGeneration;
	long double bestFitness = FLT_MAX;

	for (size_t i = 0; i < populationSize; i++) {
		vector<Individual> parents;
		parents.push_back(individuals[i]); // target
		parents.push_back(crossover->getTrial(&individuals, i)); // trial
		Individual child = crossover->performCrossover(&parents);
		long double childFitness = ((BestIndividual*)bestIndividual)->compareAndSet(&child, startCellIndex, endCellIndex);
		long double parentFitness = ((BestIndividual*)bestIndividual)->compareAndSet(&parents[0], startCellIndex, endCellIndex);

		if (childFitness < parentFitness) {
			newGeneration.push_back(child);

			if (childFitness < bestFitness) {
				bestFitness = childFitness;
			}
		}
		else {
			newGeneration.push_back(parents[0]);

			if (parentFitness < bestFitness) {
				bestFitness = parentFitness;
			}
		}
	}

	individuals = newGeneration;

	return bestFitness;
}

File: submission/12017800/code/06_decompose_coop/DifferentialEvolution.java
import java.util.ArrayList;
import java.util.List;

public class DifferentialEvolution implements Runnable {
	private Crossover crossover;
	private Mutation mutation;
	private OptimizationFunction optimizationFunction;
	private List<BestIndividual> mainPopulation;
	private List<Integer> indices = null;
	private int iterationCounter = 0;

	public DifferentialEvolution(Crossover crossover, Mutation mutation, OptimizationFunction optimizationFunction, List<BestIndividual> mainPopulation) {
		this.crossover = crossover;
		this.mutation = mutation;
		this.optimizationFunction = optimizationFunction;
		this.mainPopulation = mainPopulation;
	}

	public void setIndices(List<Integer> indices) {
		this.indices = indices;
	}

	public void setIterationCounter(int iterationCounter) {
		this.iterationCounter = iterationCounter;
	}

	public void run() {
		try {
			double bestFitness = Double.MAX_VALUE;
			List<Individual> population = new ArrayList<>();

			for(; iterationCounter >= 0; iterationCounter--) {
				for (int i = 0; i < mainPopulation.size(); i++) {
					List<Cell> cells = new ArrayList<>();

					for (int j = 0; j < indices.size(); j++) {
						int index = indices.get(j);
						cells.add(mainPopulation.get(i).getCell(index).clone());
					}

					population.add(new Individual(cells));
				}
				
				createNextGeneration(population);
			}
			// indices = null;
		}
		catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
	}

	public void createNextGeneration(List<Individual> population) {
		for (int i = 0; i < mainPopulation.size(); i++) {
			List<Individual> parents = new ArrayList<>();
			parents.add(population.get(i)); // target
			parents.add(mutation.getTrial(population, i)); // trial
			Individual child = crossover.performCrossover(parents);
			mainPopulation.get(i).compareAndSet(child, indices);
		}
	}

}